File: src/wasm/console.ts
import type { FnConsoleSetPanicHook } from '../../target/bridge-console';

import { LogLevel } from '../../target/bridge-console';
export { LogLevel };

export class WasmConsole {

    private memory?: WebAssembly.Memory;

    public init(wasm: WebAssembly.Exports) {
        this.memory = wasm.memory as WebAssembly.Memory;
        (wasm.console_set_panic_hook as FnConsoleSetPanicHook)();
    }

    public imports = {
        console_message: (level: LogLevel, ptr: number, len: number) => {
            if (this.memory == null) return;

            const chars = new Uint16Array(this.memory.buffer, ptr, len);
            const msg = String.fromCharCode(...chars);

            switch (level) {
                case LogLevel.Error:
                    console.error(msg); break;
                case LogLevel.Warn:
                    console.warn(msg); break;
                case LogLevel.Info:
                    console.info(msg); break;
                case LogLevel.Debug:
                    console.debug(msg); break;
                case LogLevel.Log:
                default:
                    console.log(msg); break;
            }
        },
    };

}

File: wasmple_console/src/console/console.rs
enum LogLevel {
    LOG,
    DEBUG,
    INFO,
    WARN,
    ERROR,
}

fn console_message_with_loglevel(level: LogLevel, msg: String) {
    let utf16: Vec<u16> = msg.encode_utf16().collect();
    unsafe {
        super::import::console_message(level as u8, utf16.as_ptr(), utf16.len());
    }
}

#[allow(dead_code)]
pub fn log(msg: String) {
    console_message_with_loglevel(LogLevel::LOG, msg);
}

#[allow(dead_code)]
pub fn debug(msg: String) {
    console_message_with_loglevel(LogLevel::DEBUG, msg);
}

#[allow(dead_code)]
pub fn info(msg: String) {
    console_message_with_loglevel(LogLevel::INFO, msg);
}

#[allow(dead_code)]
pub fn warn(msg: String) {
    console_message_with_loglevel(LogLevel::WARN, msg);
}

#[allow(dead_code)]
pub fn error(msg: String) {
    console_message_with_loglevel(LogLevel::ERROR, msg);
}
